### PR TITLE
Refactor the business logic that inserts calls to autoDestroy

### PR DESCRIPTION
**Pull request by noakesmichael**
In broad strokes the business logic is the same as before and the memory leak
statistics are largely unchanged.

The old implementation had three phases; two phases to prune auto-destroy flags and then a final 
phase to walk every block-stmt, independently and in arbitrary order, and determine which variables
are live when control leaves the block.

The new implementation has two phases; it retains one of the prune steps and then walks
functions independently and in arbitrary order.  For each function the nested block statements
are processed by top-down recursion.

The original code was a little hard to follow and had the potential to leak for functions with
multiple returns; one of the two original prune functions removed the auto-destroy flag on
every variable that might be returned even though only one variable would be returned for
a given call.

This PR has reapplied the new business logic in a handful of commits and in a top down
manner that, I hope, makes it easier to follow the important elements of the overall business
logic.  The code does not work until nearly all of the commits are applied.

This has been compiled with/without CHPL_DEVELOPER on clang and gcc and has been
tested on linux64 and gasnet.  I have also inspected the memory leak reports.  The new code
fixes a few leaks but has not had a major impact; neither for better or worse.

### CODE REVIEW

File: compiler/resolution/Makefile.share [hunk 1]
```diff
@@ -1,29 +1,30 @@
...
-	callDestructors.cpp \
-	callInfo.cpp \
-	cullOverReferences.cpp \
-	implementForallIntents.cpp \
-	generics.cpp \
-	functionResolution.cpp \
-	lowerIterators.cpp \
+RESOLUTION_SRCS =                                    \
+        addAutoDestroyCalls.cpp                      \
+	caches.cpp                                   \
```
Comment (ronawho) [inline compiler/resolution/Makefile.share]: Not sure if you fixed this later, but I think these should remain tabs instead of spaces.

Comment (noakesmichael) [inline compiler/resolution/Makefile.share]: Thanks. I update the new entry to be consistent with the other entries.
